### ISSUE TITLE
recursive combine of the 'pl.when(  ).then( ).otherwise(  )' evalute error

### ISSUE THREAD

**Issue by huacheng** [pola-rs/polars#5500]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

When i was running the test code below for times, it will print the Correct result and Wrong result randomly,
some times tobe the Correct, other times tobe the Wrong.

### Reproducible example

```python
import polars as pl
df = pl.DataFrame({
    'x': [1,2,3,4,5,6,7,8,9, 
          11,2,13,4,15,6,None,None,19], 
    'y': [0,0,0,0,0,0,0,0,0, 
          1,1,1,1,1,1,1,1,1]
})

def recursive_divide(quantile_col:str, divide_group_num:int=3, group_idx:int=1):
    if group_idx < divide_group_num:
        ####
        ge_quantile = (group_idx-1) / divide_group_num
        lt_quantile = group_idx / divide_group_num
        ####
        is_ge_expr = pl.col(quantile_col) >= pl.col(quantile_col).quantile(ge_quantile)
        is_lt_expr = pl.col(quantile_col) < pl.col(quantile_col).quantile(lt_quantile)
        #### 
        if_then_expr = (
            pl.when(
                is_ge_expr & is_lt_expr
            ).then(
                pl.lit(group_idx)
            ).otherwise(
                recursive_divide(quantile_col, divide_group_num, (group_idx + 1))
            )
        )
        print(ge_quantile, lt_quantile)
        return if_then_expr
    else:
        ####
        ge_quantile = (group_idx-1) / divide_group_num
        le_quantile = group_idx / divide_group_num
        #### 
        is_ge_expr = pl.col(quantile_col) >= pl.col(quantile_col).quantile(ge_quantile)
        is_le_expr = pl.col(quantile_col) <= pl.col(quantile_col).quantile(le_quantile)
        #### 
        if_then_expr = (
            pl.when(
                is_ge_expr & is_le_expr
            ).then(
                pl.lit(group_idx)
            ).otherwise(
                pl.lit(None)
            )
        )
        print(ge_quantile, le_quantile)
        return if_then_expr
    ####
####

####
df = df.with_columns([
    (recursive_divide('x', 3)).over(['y']).alias('divide_quantile1')
])

print(df)
```

### Expected behavior

Correct result:
_idx_, x, y, divide_quantile1
0, 1, 0, 1
1, 2, 0, 1
2, 3, 0, 1
3, 4, 0, 2
4, 5, 0, 2
5, 6, 0, 2
6, 7, 0, 3
7, 8, 0, 3
8, 9, 0, 3
9, 11, 1, 2
10, 2, 1, 1
11, 13, 1, 3
12, 4, 1, 1
13, 15, 1, 3
14, 6, 1, 2
15, None, 1, None
16, None, 1, None
17, 19, 1, 3

Wrong Result
_idx_, x, y, divide_quantile1
0, 1, 0, 2
1, 2, 0, 1
2, 3, 0, 3
3, 4, 0, 1
4, 5, 0, 3
5, 6, 0, 2
6, 7, 0, None
7, 8, 0, None
8, 9, 0, 3
9, 11, 1, 1
10, 2, 1, 1
11, 13, 1, 1
12, 4, 1, 2
13, 15, 1, 2
14, 6, 1, 2
15, None, 1, 3
16, None, 1, 3
17, 19, 1, 3

### Installed versions

<details>

```
---Version info---
Polars: 0.14.27
Index type: UInt32
Platform: Windows-10-10.0.17763-SP0
Python: 3.9.13 (tags/v3.9.13:6de2ca5, May 17 2022, 16:36:42) [MSC v.1929 64 bit (AMD64)]
---Optional dependencies---
pyarrow: 6.0.1
pandas: 1.4.1
numpy: 1.22.2
fsspec: 2022.8.2
connectorx: 0.3.0
xlsx2csv: 0.8
matplotlib: 3.5.1
```

</details>

**Comment by ritchie46** [comment]
The bug can be expressed much simpler. 

```python
df.with_columns([
    pl.when(pl.col("x") >= pl.col("x").quantile(0.1)).then(1).otherwise(None).over("y").alias("foo")
])
```

**Comment by huacheng** [comment]
thank you @ritchie46, the bug has been fixed by version 0.14.29
